src/crab/ebpf_domain.cpp

```diff
@@ -1767,11 +1762,75 @@ void ebpf_domain_t::operator()(const Bin& bin) {
                         if (!bin.is64) {
                             input &= UINT32_MAX;
                         }
-                        uint64_t output = (int64_t)(input >> imm);
+                        // 0x8000000000000000 will throw an exception if passed to m_inv.set
+                        // as an int64_t interval per arsh64.data conformance test.
+                        // But mod64.data passes with an int64_t which passes 0xb185843600000000
+                        // as does stxdw.data which passes 0x8877665500000000, but that is because
+                        // they don't have relationships.  In the future we need a way of handling
+                        // relationships between the full range of uint64_t.
+                        uint64_t output = input << imm;
                         m_inv.set(dst.value, crab::interval_t{number_t{output}});
                         break;
                     }
+                } else if (interval.finite_size()) {
+                    number_t lb = interval.lb().number().value();
+                    number_t ub = interval.ub().number().value();
+                    if (bin.is64) {
+                        uint64_t lb_n = lb.cast_to_uint64() << imm;
+                        uint64_t ub_n = ub.cast_to_uint64() << imm;
+                        m_inv.set(dst.value, crab::interval_t{lb_n, ub_n});
+                    } else {
+                        number_t lb_w = lb.cast_to_finite_width(finite_width);
+                        number_t ub_w = ub.cast_to_finite_width(finite_width);
+                        uint64_t lb_n = lb_w.cast_to_uint32() << imm;
+                        uint64_t ub_n = ub_w.cast_to_uint32() << imm;
+                        if (lb_n > ub_n) {
+                            // Range is an int32_t, so use the full uint32_t range.
+                            m_inv.set(dst.value, crab::interval_t{0, UINT32_MAX});
+                        } else {
+                            // Range is contiguous as a uint32_t.
+                            m_inv.set(dst.value, crab::interval_t{lb_n, ub_n});
+                        }
+                    }
+                    break;
+                } else if (!bin.is64) {
+                    m_inv.set(dst.value, crab::interval_t{0, UINT32_MAX});
```

Comment:
Untested branch.

Reply:
Actually this can't happen due to line 1713, so removed.